compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/scene/ComposeSceneContext.skiko.kt

```
@@ -72,12 +72,7 @@ interface ComposeSceneContext {
         throw IllegalStateException()
     }
 
-    companion object {
-        /**
-         * Represents an empty implementation of [ComposeSceneContext] and used to provide
-         * a default value.
-         */
-        val Empty = object : ComposeSceneContext {
-        }
+    class Base : ComposeSceneContext {
```

The semantics of this class is not to be a base for others. The semantic of this class is to be maximum behavior-less, "empty" context/platform. And it is correct to use just it without adding an additional behavior.

I would at least keep the name "Empty", or think about changing it to something else.

---

> think about changing it to something else.

Alternative is Blank

---

I don't know... I don't see a huge difference between the two here. Now that it is a class that you typically subclass rather than use directly, I think `Base` is more appropriate.

What do you think about `EmptyBase`?

---

Also, it's not actually "empty", or we wouldn't be here. An "empty" implementation would ignore input mode change requests.

---

> I don't see a huge difference between the two here

I see difference in semantics.

> Now that it is a class that you typically subclass rather than use directly, I think Base is more appropriate.

Delegation to Empty (or subclassing) wasn't there initially - you either used the Empty, or implement every method in the context. It was only later we added delegation, but I am still not sure if it was right (we can miss new methods). Empty was created only as the default for `ComposeScene` (which was public).

Besides semantics, there is a code style - delegating to a real abstraction ("Empty") is better than subclassing a not complete abstraction ("Base").

>  An "empty" implementation would ignore input mode change requests.

It also has to maintain the contract "if you called requestInputMode method, inputMode field has to be changed"

---

> I see difference in semantics.

By "semantics" you mean "how it's meant to be used", but now it is actually (almost?) always used as a superclass.

> Delegation to Empty (or subclassing) wasn't there initially - you either used the Empty, or implement every method in the context

But again, subclassing is how it's used now.

> It also has to maintain the contract "if you called requestInputMode method, inputMode field has to be changed"

Sure, but that makes it not "empty". It's more "default" than "empty".

Also, other types under PlatformContext also have certain contracts, but their "empty" implementations don't fulfill them.

---

> By "semantics" you mean "how it's meant to be used"

No, I mean what this class provides. "Empty" provides a behaviorless but complete context. "Base" by it's name provides something incomplete that needs to be extended.

> but now it is actually (almost?) always used as a superclass.

It can be used as a superclass, but we need to keep the initial semantics. It is still correct just take it as is and use.

> It's more "default" than "empty".

In my eyes Empty or Blank tells more, but Default also works.

> Also, other types under PlatformContext also have certain contracts, but their "empty" implementations don't fulfill them.

If there is such a class, it isn't by the design. Do you have an example?

---

> If there is such a class, it isn't by the design. Do you have an example?

Calling `FocusManager.clearFocus` should clear focus but `EmptyFocusManager.clearFocus` doesn't.
Calling `TextToolbar.showMenu` should cause `TextToolbar.status` to change to `Shown`, but `EmptyTextToolbar.showMenu` doesn't.

---

> In my eyes Empty or Blank tells more, but Default also works.

Renamed it to `Default`.

---

We need to rename other classes/objects that start with Empty then (EmptyPlatformTextInputService, etc)

---

> We need to rename other classes/objects that start with Empty then (EmptyPlatformTextInputService, etc)

I disagree, because they truly are "empty". They can't be "used" as-is as a default implementation of their interfaces. The current implementation of `EmptyPlatformTextInputService` or `EmptyFocusManager` can't be used as a "default"; it would be extremely misleading.

I did rename `EmptyViewConfiguration` to `DefaultViewConfiguration`, because for that it does make sense .

---

and on that note, `PlatformContext.Default` (previously `Base`) feels like what you defined "base" to be (even though I disagree that that should be the definition). It's not a complete implementation that can be used as-is without subclassing. It can be used like that in tests, where you're testing a narrow feature, but not in a real implementation, because it's not actually implementing all the contracts (as I showed a few messages back).

---

> I disagree, because they truly are "empty". They can't be "used" as-is as a default implementation of their interfaces.

Then I still incline to return it to back to Empty. Or rename it to Blank if it is better for you. The semantics of all these classes is the same and it is to provide behavior-less abstractions. And in my eyes even if it is called Empty it still can be stateful, so I don't see any contradiction.

>  It's not a complete implementation that can be used as-is without subclassing. It can be used like that in tests, where you're testing a narrow feature, but not in a real implementation

It is not a real implementation by design, but it is complete and should and can be used where you need a minimal not real implementation. Either in tests, or even in real application in background to simulate something.

> feels like what you defined "base" to be

It may looks this way, but it may also looks the way I describe - to not be the base. And the only way to know the purpose is to read the docs (which we didn't add 🫠), or ask the author (I added it, Ivan refactored). We may change the purpose to be it as a base, but we lose the purpose of it to be the behaviorless context. And I am against it, because I see that we still need this as standalone behaviorless context in some cases.

So, the main purpose is to use it this way:
```
fun CanvasLayersComposeScene(
    ...
    platformContext: PlatformContext = PlatformContext.Default(),
    ...
)
```

And this way was is only a side-effect, but still a valid one:
```
class MyContext : PlatformContext by PlatformContext.Default()
```

> (as I showed a few messages back).

The fact that they don't implement the contract is not by the design. We should fix that (but not here, it isn't the purpose of this PR):

- `PlatformContext.FocusManager` is for the owner focus manager. If there is no owner, it should be null, and we shouldn't have Empty at all.
- `TextToolbar.showMenu` should either change the state, or we should remove `EmptyTextToolbar` and make it nullable

---

Alternatives to Empty/Blank:
- Virtual
- InMemory

These names probably better describe why I am against treating it as a base.

---

I renamed it back to Empty.